### PR TITLE
dRPA_k_mGW

### PR DESCRIPTION
**Pull request by mkakcl**
Adding dRPA functionality to the PBC code. Similar overall structure to the dRPA molecular code, with influence from the PBC-dTDA code. Some functionality is share with the dTDA code, such as the convolutions. Within these shared functions, some optimisations were added.

Further optimisations throughout other parts of the code in line with #43 (the `einsum` ellipsis changes).

Added tests to ensure PBC-dRPA runs smoothly.

The `ints.py` for PBC are reverted back to previous version, see changes occurring in #45.

### CODE REVIEW

File: momentGW/energy.py
Comment (obackhouse) [inline momentGW/energy.py:11]: Use `einsum` -- it's much clearer. We only need to worry about the optimisation of rate-limiting calls.

File: momentGW/energy.py
Comment (obackhouse) [inline momentGW/energy.py:59]: Revert to `einsum`

File: momentGW/pbc/kpts.py
Comment (obackhouse) [inline momentGW/pbc/kpts.py:360]: what's going on with these changes?

File: momentGW/pbc/rpa.py
Comment (obackhouse) [inline momentGW/pbc/rpa.py:66]: Improve docstrings for integrals, i.e. "keys should be the pair of k-point indices, and values are the ... at that k-point pair"

File: momentGW/pbc/rpa.py
Comment (obackhouse) [inline momentGW/pbc/rpa.py:90]: we certainly shouldn't be storing these as class attributes -- it stops the garbage collector dealing with them. Follow the same standards as the other polarizability classes, i.e. pass them as arguments.

File: momentGW/pbc/rpa.py
Comment (obackhouse) [inline momentGW/pbc/rpa.py:92]: we can inherit this from the molecular dRPA -- just need to move the quadrature error reporting to a separate function, and redeclare it for k-space

File: momentGW/pbc/rpa.py
Comment (obackhouse) [inline momentGW/pbc/rpa.py:139]: what's the status of MPI here? there are MPI enabled loops but missing reductions etc.

File: momentGW/pbc/rpa.py
Comment (obackhouse) [inline momentGW/pbc/rpa.py:160]: clean up these integral docstrings

File: momentGW/pbc/rpa.py
Comment (obackhouse) [inline momentGW/pbc/rpa.py:233]: typo

File: momentGW/pbc/rpa.py
Comment (obackhouse) [inline momentGW/pbc/rpa.py:240]: MPI?

File: momentGW/pbc/rpa.py
Comment (obackhouse) [inline momentGW/pbc/rpa.py:271]: MPI?

File: momentGW/pbc/rpa.py
Comment (obackhouse) [inline momentGW/pbc/rpa.py:436]: docstrings

File: momentGW/pbc/rpa.py
Comment (obackhouse) [inline momentGW/pbc/rpa.py:478]: remove, will be inherited

File: momentGW/pbc/rpa.py
Comment (obackhouse) [inline momentGW/pbc/rpa.py:483]: remove, will be inherited

File: momentGW/pbc/rpa.py
Comment (obackhouse) [inline momentGW/pbc/rpa.py:491]: remove, will be inherited

File: momentGW/pbc/rpa.py
Comment (obackhouse) [inline momentGW/pbc/rpa.py:496]: remove, will be inherited

File: momentGW/pbc/rpa.py
Comment (obackhouse) [inline momentGW/pbc/rpa.py:15]: do we want the MRO to be in this order? probably favour inheritence from `dTDA` over `MoldRPA` i would think

File: momentGW/pbc/tda.py
Comment (obackhouse) [inline momentGW/pbc/tda.py:157]: `eo * fh`

File: momentGW/pbc/tda.py
Comment (obackhouse) [inline momentGW/pbc/tda.py:162]: `ev * fp`

File: momentGW/pbc/tda.py
Comment (obackhouse) [reply momentGW/pbc/tda.py:157]: no need to initialise `tmp_to` really

File: momentGW/pbc/tda.py
Comment (obackhouse) [reply momentGW/pbc/tda.py:162]: also as above

File: momentGW/pbc/tda.py
Comment (obackhouse) [inline momentGW/pbc/tda.py:229]: probs prefer an `einsum`

File: momentGW/util.py
Comment (obackhouse) [inline momentGW/util.py:275]: don't think we need to force the contiguity of the output here -- i can take a look at this function

**Review by obackhouse — CHANGES_REQUESTED** [review]
Added some comments, need to restructure the arguments and inheritance, we can do a few iterations of reviews then get this merged

File: momentGW/pbc/rpa.py
Comment (mkakcl) [reply momentGW/pbc/rpa.py]: With changes to the inheritance of objects like Liad, I don't think this will be possible. Can move them into the integrate class, but then won't be the same as the molecular dRPA.

**Comment by mkakcl** [comment]
Still doing some MPI testing, so will respond to those comments once that is finished.

File: momentGW/pbc/kpts.py
Comment (mkakcl) [reply momentGW/pbc/kpts.py]: These changes where due to situations where I would get say `[-0,-0,-0]` rather than `[0,0,0]`. Am guessing this was caused by something else as  I cannot replicate the issue any more. Have reverted the code to the previous version.

**Comment by obackhouse** [comment]
I've moved the calculation of the intermediates to separate functions and also made them more like the molecular code in terms of what is passed as function arguments. I can see that you were trying to avoid multiple calculation of `Liad` and `d`, but to be honest this is not going to be a bottleneck so some repetition is fine.

I actually think we may just want to completely do away with `Liad` and `Liadinv` for solids. I don't think the benefit of those intermediates will at all outweight the massive memory dependence, so it doesn't make any sense to keep them. Intermediates in PBC calculations are often only worth the memory overhead if you can fuse your k-point loops and build the intermediates just for that k-point pair -- maybe we are missing some k-point loop fusion opportunities somewhere.

**Comment by obackhouse** [comment]
LGTM -- i think we have some room for improvement with future refactoring, particularly with 

> I actually think we may just want to completely do away with Liad and Liadinv for solids. I don't think the benefit of those intermediates will at all outweight the massive memory dependence, so it doesn't make any sense to keep them. Intermediates in PBC calculations are often only worth the memory overhead if you can fuse your k-point loops and build the intermediates just for that k-point pair -- maybe we are missing some k-point loop fusion opportunities somewhere.

but let's leave that for a future PR and get this merged, great work man.

UHF next? :laughing:

**Review by obackhouse — APPROVED** [review]

**Comment by obackhouse** [comment]
oh and @mkakcl can you remind me the status on what effect merging this has on #45 ?

**Comment by mkakcl** [comment]
Waiting on the merge of #45 with the change on `ao_loc` before merging this.

**Pull request merged**
